### ISSUE TITLE
ignore_index does not work as described with nn.CrossEntropyLoss

### ISSUE THREAD

**Issue by wgale** [pytorch/pytorch#63004]
## 🐛 Bug

Ignored targets/indices impact the output of nn.CrossEntropyLoss/torch.nn.functional.cross_entropy, have gradients associated with them, and impact the gradients of all other classes. This is a result of "ignore_index" only being used with nll_loss in the underlying implementation and not being used in the log_softmax. The log_softmax is over all the classes, including any classes that the loss function has been instructed to ignore, which means the ignored targets appear in the denominator for all the other classes.

If ignore_index is negative then the loss works as intended, but if the ignore_index is a valid index then it will not be ignored and it will impact the input gradient.
A similar issue exists with "weight" and zero-weight targets will also impact the gradient/loss - this is unavoidable but it probably should be clearly stated in the documentation as it could result in undesirable behavior. 

## To Reproduce

Repo case showing the impact on the loss (l2 should produce the same loss for all three examples as it ignores index 0):
```python
>>> import torch
>>> x1 = torch.as_tensor([[1., 1., 1.], [1., 1., 1.]])
>>> x2 = torch.as_tensor([[0., 1., 1.], [0., 1., 1.]])
>>> x3 = torch.as_tensor([[float("-inf"), 1., 1.], [float("-inf"), 1., 1.]])
>>> y = torch.as_tensor([1, 2])
>>> l1 = torch.nn.CrossEntropyLoss(reduction="none")
>>> l2 = torch.nn.CrossEntropyLoss(ignore_index=0, reduction="none")
>>> l3 = torch.nn.CrossEntropyLoss(weight=torch.as_tensor([0., 1., 1.]), reduction="none")
>>> l1(x1,y)
tensor([1.0986, 1.0986])
>>> l1(x2,y)
tensor([0.8620, 0.8620])
>>> l1(x3,y)
tensor([0.6931, 0.6931])
>>> l2(x1,y)
tensor([1.0986, 1.0986])
>>> l2(x2,y)
tensor([0.8620, 0.8620])
>>> l2(x3,y)
tensor([0.6931, 0.6931])
>>> l3(x1,y)
tensor([1.0986, 1.0986])
>>> l3(x2,y)
tensor([0.8620, 0.8620])
>>> l3(x3,y)
tensor([0.6931, 0.6931])
```
Repo case showing that "ignore_index" has a gradient associated with it (the gradient for the first column should be 0):
```python
>>> x = y = torch.as_tensor([1,2,3])
>>> m = torch.nn.Embedding(4,4,padding_idx=0)
>>> crit = torch.nn.CrossEntropyLoss(ignore_index=0)
>>> y_hat = m(x)
>>> loss = crit(y_hat,y)
>>> loss.backward()
>>> m.weight.grad
tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0436, -0.1709,  0.0204,  0.1069],
        [ 0.0729,  0.1171, -0.2597,  0.0698],
        [ 0.1109,  0.0582,  0.0490, -0.2181]])
```

## Expected behavior
Expected outputs for the first repo case showing the impact on the loss:
```python
>>> import torch
>>> x1 = torch.as_tensor([[1., 1., 1.], [1., 1., 1.]])
>>> x2 = torch.as_tensor([[0., 1., 1.], [0., 1., 1.]])
>>> x3 = torch.as_tensor([[float("-inf"), 1., 1.], [float("-inf"), 1., 1.]])
>>> y = torch.as_tensor([1, 2])
>>> l1 = torch.nn.CrossEntropyLoss(reduction="none")
>>> l2 = torch.nn.CrossEntropyLoss(ignore_index=0, reduction="none")
>>> l3 = torch.nn.CrossEntropyLoss(weight=torch.as_tensor([0., 1., 1.]), reduction="none")
>>> l1(x1,y)
tensor([1.0986, 1.0986])
>>> l1(x2,y)
tensor([0.8620, 0.8620])
>>> l1(x3,y)
tensor([0.6931, 0.6931])
>>> l2(x1,y)
tensor([0.6931, 0.6931])
>>> l2(x2,y)
tensor([0.6931, 0.6931])
>>> l2(x3,y)
tensor([0.6931, 0.6931])
>>> l3(x1,y)
tensor([1.0986, 1.0986])
>>> l3(x2,y)
tensor([0.8620, 0.8620])
>>> l3(x3,y)
tensor([0.6931, 0.6931])
```
Repo case showing that "ignore_index" has a gradient associated with it (the gradient for the first column should be 0):
```python
>>> x = y = torch.as_tensor([1,2,3])
>>> m = torch.nn.Embedding(4,4,padding_idx=0)
>>> crit = torch.nn.CrossEntropyLoss(ignore_index=0)
>>> y_hat = m(x)
>>> loss = crit(y_hat,y)
>>> loss.backward()
>>> m.weight.grad
tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000, -0.1709,  0.0204,  0.1069],
        [ 0.0000,  0.1171, -0.2597,  0.0698],
        [ 0.0000,  0.0582,  0.0490, -0.2181]])
```

## Environment
N/A

## Additional context
I'm currently working around this issue by using a boolean ignore mask (vector of length num_classes): 
```python
class CrossEntropyLoss(torch.nn.CrossEntropyLoss):
    def __init__(self, weight=None, ignore_index=-100, ignore_mask=None, reduction='mean'):
        self.ignore_mask = ignore_mask
        if ignore_mask is not None and weight is None:
            weight = ignore_mask.eq(False).float()
        super().__init__(weight=weight, ignore_index=ignore_index, reduction=reduction)

    def forward(self, input: Tensor, target: Tensor) -> Tensor:
        masked = input.clone()
        masked[:, self.ignore_mask] = float("-inf")
```
There's some previous discussion of this issue in #25166, but it was closed with a workaround.

**Comment by jbschlosser** [comment]
Thanks @wgale for the detailed report! The current semantics of `ignore_index` only ignore the specified class index when encountered within the *target*:

```python
import torch

torch.manual_seed(2)

input = torch.randn(5, 3, requires_grad=True)
target = torch.tensor([2, 0, 1, 2, 1])
crit = torch.nn.CrossEntropyLoss(ignore_index=1)
crit(input, target).backward()
print(input.grad)
```
```
tensor([[ 0.1641,  0.0887, -0.2528],  # target = 2 for this item
        [-0.3061,  0.2579,  0.0482],  # target = 0 for this item
        [ 0.0000,  0.0000,  0.0000],  # target = 1 for this item
        [ 0.0189,  0.2009, -0.2198],  # target = 2 for this item
        [ 0.0000,  0.0000,  0.0000]])  # target = 1 for this item
```
Note that input gradients for indices 2 and 4 are zero, corresponding with the indices where `target=ignore_index=1`.

It seems that for your use case, you need input logits for class of `ignore_index` to be ignored as well / not take part in the softmax - is that correct? (related [forum post](https://discuss.pytorch.org/t/ignore-index-in-the-cross-entropy-loss/25006)).

**Comment by wgale** [comment]
Thanks @jbschlosser, that makes sense. It looks like there's already an existing issue about this in #18206.

It sounds like the bug is more of a documentation issue, although adding the functionality to ignore indices could also be useful. Is this something that might be able to be included in PyTorch?

Thanks for you interest @syedaasim033. The CrossEntropyLoss contains the changes needed to ignore indices from the loss, although it's and unsafe and inefficient as I simplified it to make it clearer.

**Comment by jbschlosser** [comment]
> It sounds like the bug is more of a documentation issue, although adding the functionality to ignore indices could also be useful. Is this something that might be able to be included in PyTorch?

Possibly, although I think it'd require some kind of masked_log_softmax to implement. cc @cpuhrsch
Might be worth opening a new feature request linking this issue and the other related ones for identifying interest & gathering design feedback.

Btw, the backward pass for `Embedding` won't apply gradient updates to the item at `padding_idx`. Given that, do you still need the new behavior?

**Comment by wgale** [comment]
> Possibly, although I think it'd require some kind of masked_log_softmax to implement. cc cpuhrsch
> Might be worth opening a new feature request linking this issue and the other related ones for identifying interest & gathering design feedback.

That would be great - thank you!

> Btw, the backward pass for `Embedding` won't apply gradient updates to the item at `padding_idx`. Given that, do you still need the new behavior?

Yes, I think the new behavior is important despite the embedding layer handling the padding_idx. There are a couple of use cases where I think this might impact sequence modeling:
- Tied embedding and softmax weights  - this might not be an issue depending on how the gradient updates are applied
- Handling null tokens (it's common to pad the vocabulary to be mod 8 when using FP16/TensorCores, the null tokens should be excluded from the loss)

Also, from what I understand, the padding class will still appear in the denominator for the other classes in the CE loss even though the gradients for the padding class itself won't updated. This could impose an additional, unintended, constraint on the model.

**Comment by blueardour** [comment]
Hi, is there any plan to add the 'ignore_index' rather than 'ignore_target' feature?  I also met an issue which could be solved by 'ignore_index' on a classification task.
